### PR TITLE
Task/inline suggestion

### PR DESCRIPTION
**Pull request by dimacodota**
add inline suggestion capability

### CODE REVIEW

File: src/inlineSuggestions/inlineSuggestionsLifecycle.ts [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+} from "vscode";
+import acceptInlineSuggestion from "./acceptInlineSuggestion";
+import clearInlineSuggestionsState from "./clearDecoration";
+import { getNextSuggestion, getPrevSuggestion } from "./inlineSuggestionState";
+import setInlineSuggestion from "./setInlineSuggestion";
+import textListener from "./textListener";
+
+export const decorationType = window.createTextEditorDecorationType({});
+
+export default async function inlineSuggestionsLifecycle(
```
Comment (dimacodota) [inline src/inlineSuggestions/inlineSuggestionsLifecycle.ts:17]: this is the "main" function

File: src/extension.ts [hunk 2]
```diff
@@ -91,13 +93,19 @@ async function backgroundInit(context: vscode.ExtensionContext) {
...
   void executeStartupActions();
-  vscode.languages.registerCompletionItemProvider(
-    { pattern: "**" },
-    {
-      provideCompletionItems,
-    },
-    ...COMPLETION_TRIGGERS
-  );
+  const suggestionsMode = getSuggestionMode();
+  if (suggestionsMode === SuggestionsMode.INLINE) {
```
Comment (bilucodota) [inline src/extension.ts:97]: consider use `switch case`

File: src/getSuggestionMode.ts [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import { Capability, isCapabilityEnabled } from "./capabilities";
```
Comment (bilucodota) [inline src/getSuggestionMode.ts:1]: consider nest in `src/capabilities `

File: src/inlineSuggestions/inlineSuggestionsLifecycle.ts (hunk 1 above)
Comment (bilucodota) [reply src/inlineSuggestions/inlineSuggestionsLifecycle.ts:17]: consider changing to `registerHandlers`

File: src/inlineSuggestions/inlineSuggestionState.ts [hunk 4]
```diff
@@ -0,0 +1,51 @@
+import { commands } from "vscode";
+import { AutocompleteResult, ResultEntry } from "../binary/requests/requests";
+import { rotate } from "./rotate";
+
+let autocompleteResult: AutocompleteResult | undefined | null;
+
+let iterator = rotate(0);
+
+export async function setSuggestionsState(
```
Comment (bilucodota) [inline src/inlineSuggestions/inlineSuggestionState.ts:9]: change state cross tabs

File: src/inlineSuggestions/setInlineSuggestion.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    prefix || ""
+  );
+  return suggestedHint.replace(new RegExp(`${escapeRegExp(suffix)}$`), "");
+}
+
+function clearPrefixFromSuggestion(currentCompletion: string, prefix: string) {
+  return currentCompletion?.replace(prefix, "");
+}
+
+function escapeRegExp(value: string) {
```
Comment (bilucodota) [inline src/inlineSuggestions/setInlineSuggestion.ts:58]: move to `util`

File: src/inlineSuggestions/setInlineSuggestion.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+  prefix: string | undefined
+): string {
+  const suffix = document.getText(
+    new Range(position, document.lineAt(position.line).range.end)
+  );
+  const suggestedHint = clearPrefixFromSuggestion(
+    newSuggestion?.new_prefix || "",
+    prefix || ""
+  );
+  return suggestedHint.replace(new RegExp(`${escapeRegExp(suffix)}$`), "");
```
Comment (bilucodota) [inline src/inlineSuggestions/setInlineSuggestion.ts:51]: move to util - `replaceSuffix` or something similar

File: src/inlineSuggestions/setInlineSuggestion.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+function showInlineDecoration(position: Position, suggestion: string): void {
+  const suggestionDecoration: DecorationOptions = {
+    renderOptions: {
+      after: {
+        color: "gray",
+        contentText: suggestion,
+      },
+    },
+    range: new Range(
+      new Position(position.line, position.character),
```
Comment (bilucodota) [inline src/inlineSuggestions/setInlineSuggestion.ts:71]: pass `position`

File: src/inlineSuggestions/setInlineSuggestion.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      new Position(position.line, position.character)
+    ),
+  };
+  const hoverDecoration: DecorationOptions = {
+    hoverMessage: hoverPopup,
+    range: new Range(
+      new Position(position.line, position.character),
```
Comment (bilucodota) [inline src/inlineSuggestions/setInlineSuggestion.ts:78]: pass `position`

File: src/inlineSuggestions/setInlineSuggestion.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+      position.translate(undefined, position.character)
```
Comment (bilucodota) [inline src/inlineSuggestions/setInlineSuggestion.ts:79]: missing text length?

File: src/inlineSuggestions/hoverPopup.ts [hunk 10]
```diff
@@ -0,0 +1,11 @@
+import { MarkdownString } from "vscode";
+import { FULL_BRAND_REPRESENTATION, IS_OSX } from "../globals/consts";
+
+const space = "&nbsp;&nbsp;&nbsp;";
+const altKey = IS_OSX ? "\u2325" : "alt";
+const hoverPopupContent = `[Next \\(${altKey}\\]\\)](command:tabnine.next-inline-suggestion)${space}[Prev \\(${altKey}\\[\\)](command:tabnine.prev-inline-suggestion)${space}[Accept \\(tab\\)](command:tabnine.accept-inline-suggestion)${space}[Escape \\(esc\\)](command:tabnine.escape-inline-suggestion)${space}${FULL_BRAND_REPRESENTATION}`;
```
Comment (bilucodota) [inline src/inlineSuggestions/hoverPopup.ts:6]: move all commands to seapraate file

Comment (bilucodota) [reply src/inlineSuggestions/hoverPopup.ts:6]: **T**ab

File: src/inlineSuggestions/runCompletion.ts [hunk 11]
```diff
@@ -0,0 +1,22 @@
+import { Position, Range, TextDocument } from "vscode";
+import { autocomplete, AutocompleteResult } from "../binary/requests/requests";
+import { CHAR_LIMIT, MAX_NUM_RESULTS } from "../globals/consts";
+
+export default async function runCompletion(
```
Comment (bilucodota) [inline src/inlineSuggestions/runCompletion.ts:5]: maybe we should unify both commands in a single file/source

File: src/inlineSuggestions/acceptInlineSuggestion.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+  prefix: string,
+  oldPrefix: string
+) {
+  return new Range(
+    currentTextPosition.translate(undefined, -prefix.length),
+    currentTextPosition.translate(undefined, oldPrefix.length)
+  );
+}
+
+export function escapeTabStopSign(value: string): string {
```
Comment (bilucodota) [inline src/inlineSuggestions/acceptInlineSuggestion.ts:60]: move to `util`?

File: src/inlineSuggestions/rotate.ts [hunk 13]
```diff
@@ -0,0 +1,28 @@
+export type Iterator = {
+  next: () => number;
+  prev: () => number;
+  current: () => number;
+};
+
+export function rotate(value: number): Iterator {
```
Comment (bilucodota) [inline src/inlineSuggestions/rotate.ts:7]: move to `util`, use `%`

**Pull request merged**